### DISCUSSION TITLE
Apply custom function with Dataframes as arguments

### DISCUSSION THREAD

**Discussion by christosmak** [dask/dask#8419] [Q&A]
Hello everyone,

I have successfully parallelized my code with Dask, however I think that there is room for performance optimization on what I have done. Given that I am quite new to Dask and Python in general, I cannot quite easily wrap my head around Dask documentation in order to address the issue that I am having.

To begin with, here are some lines of code:

```python
import pandas as pd
import multiprocessing as mp
import dask.dataframe as dd

def custom_function(temp, df1, df2, df3, df4)->list:
    ### Calculations that involve 1 merge and filterings/assignments of values to the Dataframes passed
return single_Dataframe

# temp: Dataframe with 1 column named ['keys']

cluster = LocalCluster(n_workers=mp.cpu_count()-1, processes=True)
client = Client(cluster)
ddf = dd.from_pandas(temp, npartitions=mp.cpu_count()-1).reset_index().set_index('index')
result = ddf.apply(custom_func, axis=1, args=(df1, df2, df3, df4,), meta=('df','O')).compute()
client.shutdown()
```

My main issue here is that it takes too long for Dask to apply the function on temp, even on a machine with 24 processors. The same code in R (using the built-in library with parLapply) takes 1/3 of the time. I suspect that I am doing something wrong with distributing work to the Client, I have tried to use other features of Dask such as client.submit or client.scatter, but could not get any of them to work.

Could there be a more efficient way to do this using submit or scatter?

Thank you in advance.

**Comment by SultanOrazbayev** [comment]
It might be possible to get good speed-up with `.map_partitions` instead of `.apply`. The indexing could also be a bottleneck... but how to resolve it depends on the specifics of a case. There is some relevant discussion here: https://targomo.medium.com/how-we-learned-to-love-dask-and-achieved-a-40x-speedup-aa14e72d99c0
